File: frontend/src/types/calibration-run.ts
import type { ISODateString, UUID } from './common';

// §11 캘리브레이션 — 백엔드 schemas/calibration_run.py 와 정합.

export type CalibrationRunStatus =
  | 'pending'
  | 'running'
  | 'succeeded'
  | 'failed'
  | string;

/** 공간 유형 — calibration 의 soft prior 로 사용. 백엔드 SpaceTypeLiteral 과 정합. */
export type SpaceType =
  | 'cafe'
  | 'study_room'
  | 'classroom'
  | 'office'
  | 'residential'
  | 'unknown';

export interface CalibrationRun {
  id: UUID;
  status: CalibrationRunStatus;
  session_id: UUID | null;
  rf_run_id: UUID | null;
  version_id: UUID;
  /** RMSE, MAE 등 보정 결과 메트릭. 백엔드 service 가 채워 넣음. */
  error_metrics_json: Record<string, unknown>;
  /** 보정 전후 차이 시각화 heatmap (presigned URL). 없을 수 있음. */
  error_heatmap_url: string | null;
  created_at: ISODateString;
  finished_at: ISODateString | null;
}

export type MeasurementPurpose = 'calibration' | 'validation' | 'reference' | 'unknown';

export interface CalibrationEvaluationRequest {
  floor_id: UUID;
  scene_version_id: UUID;
  rf_run_id: UUID;
  measurement_session_ids: UUID[];
  method?: 'global_offset';
  split?: {
    strategy: 'purpose_or_random' | 'random';
    holdout_ratio: number;
    seed: number;
  };
  visualization?: {
    include_reference_map: boolean;
    reference_map_method: 'idw';
    rssi_min_dbm: number;
    rssi_max_dbm: number;
  };
}

export interface CalibrationEvaluationPoint {
  point_id: UUID;
  session_id: UUID;
  x_m: number;
  y_m: number;
  rssi_dbm: number;
  measurement_purpose: MeasurementPurpose;
  split: MeasurementPurpose;
  baseline_pred_dbm?: number;
  calibrated_pred_dbm?: number;
  baseline_error_db?: number;
  calibrated_error_db?: number;
  invalid_reason?: string;
}

export interface CalibrationEvaluationMap {
  label: string;
  values_dbm: number[][];
  bounds_m: { min_x: number; min_y: number; max_x: number; max_y: number };
  method?: string;
  offset_db?: number;
  metadata?: Record<string, unknown>;
}

export interface CalibrationEvaluationResponse {
  calibration_run_id: UUID;
  status: string;
  maps: {
    baseline: CalibrationEvaluationMap;
    calibrated: CalibrationEvaluationMap;
    measured_reference?: CalibrationEvaluationMap;
  };
  color_scale: { min_dbm: number; max_dbm: number };
  points: {
    calibration: CalibrationEvaluationPoint[];
    validation: CalibrationEvaluationPoint[];
    reference: CalibrationEvaluationPoint[];
    invalid: CalibrationEvaluationPoint[];
  };
  metrics: Record<string, number>;
  evaluation: Record<string, unknown>;
}

export interface CalibrationRunCreateRequest {
  session_id: UUID;
  rf_run_id: UUID;
  version_id: UUID;
  /** 공간 유형 soft prior. 미지정 시 백엔드가 'unknown' 으로 fallback. */
  space_type?: SpaceType;
}

/** §11.3 파라미터 변경 이력 — 어떤 wall/object 의 무슨 파라미터가 어떤 값으로 바뀌었는지. */
export interface ParameterUpdate {
  id: UUID;
  calibration_run_id: UUID;
  target_type: string; // 'wall' | 'object' 등
  target_id: UUID;
  param_name: string;
  old_value_json: unknown;
  new_value_json: unknown;
  created_at: ISODateString;
}

File: backend/app/schemas/rf/calibration_run.py
"""§11 Calibration DTO"""
from __future__ import annotations

from datetime import datetime
from typing import Any, Literal, Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator

from app.core.enums import normalize_job_status

# SpaceType 의 string 값들을 Literal 로 받아 Pydantic 검증 — runner 의 resolve_space_type
# 이 unknown fallback 처리하므로 빠져도 안전하지만 typo 는 여기서 차단.
SpaceTypeLiteral = Literal[
    "cafe", "study_room", "classroom", "office", "residential", "unknown"
]

MeasurementPurposeLiteral = Literal["calibration", "validation", "reference", "unknown"]


class CalibrationRunCreate(BaseModel):
    """POST /calibration-runs — 명세 §11.1"""
    model_config = ConfigDict(extra="forbid")

    session_id: UUID
    rf_run_id: UUID
    version_id: UUID
    # 공간 유형 (soft prior). 미지정 시 backend runner 가 "unknown" 으로 fallback.
    space_type: Optional[SpaceTypeLiteral] = None


class CalibrationEvaluationSplit(BaseModel):
    model_config = ConfigDict(extra="forbid")

    strategy: Literal["purpose_or_random", "random"] = "purpose_or_random"
    holdout_ratio: float = Field(default=0.3, gt=0.0, lt=1.0)
    seed: int = 42


class CalibrationEvaluationVisualization(BaseModel):
    model_config = ConfigDict(extra="forbid")

    include_reference_map: bool = True
    reference_map_method: Literal["idw"] = "idw"
    rssi_min_dbm: float = -90.0
    rssi_max_dbm: float = -30.0


class CalibrationEvaluationRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    floor_id: UUID
    scene_version_id: UUID
    rf_run_id: UUID
    measurement_session_ids: list[UUID] = Field(default_factory=list, min_length=1)
    ap_bssid: str | None = None
    method: Literal["affine_rssi_transfer", "global_offset"] = "affine_rssi_transfer"
    split: CalibrationEvaluationSplit = Field(default_factory=CalibrationEvaluationSplit)
    visualization: CalibrationEvaluationVisualization = Field(
        default_factory=CalibrationEvaluationVisualization
    )


class CalibrationEvaluationResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    calibration_run_id: UUID
    status: str
    maps: dict[str, Any]
    color_scale: dict[str, float]
    points: dict[str, list[dict[str, Any]]]
    metrics: dict[str, float]
    evaluation: dict[str, Any]


class CalibrationRunResponse(BaseModel):
    """GET /calibration-runs/{id} 응답 — 명세 §11.2

    모델 컬럼명과 명세 필드명이 다른 부분은 service 의 _to_response 에서 alias 매핑.
    """
    model_config = ConfigDict(extra="forbid")

    id: UUID
    status: str
    session_id: Optional[UUID] = None
    rf_run_id: Optional[UUID] = None
    version_id: UUID
    error_metrics_json: dict[str, Any] = Field(default_factory=dict)
    error_heatmap_url: Optional[str] = None
    created_at: datetime
    finished_at: Optional[datetime] = None

    @field_validator("status", mode="before")
    @classmethod
    def _normalize_status(cls, v: Any) -> Any:
        """내부 status(completed/queued) → API 표기(succeeded/pending). 프론트 폴링 종료 인식용."""
        if isinstance(v, str):
            return normalize_job_status(v)
        return v


class CalibrationRunUpdate(BaseModel):
    """[시스템] PATCH /calibration-runs/{id} — AI 워커가 진행 상태 갱신."""
    model_config = ConfigDict(extra="forbid")

    status: Optional[str] = None
    metrics_json: Optional[dict[str, Any]] = None
    error_message: Optional[str] = None


class ParameterUpdateCreate(BaseModel):
    """[시스템] POST /calibration-runs/{id}/parameter-updates"""
    model_config = ConfigDict(extra="forbid")

    target_type: str = Field(..., min_length=1, max_length=20)
    target_id: UUID
    param_name: str = Field(..., min_length=1, max_length=80)
    old_value_json: Any | None = None
    new_value_json: Any | None = None

    @model_validator(mode="after")
    def _at_least_one_value(self):
        # 새 값 정도는 있어야 의미 있음 (old 만 있으면 정보 부족).
        if self.new_value_json is None and self.old_value_json is None:
            raise ValueError("at least one of old_value_json/new_value_json required")
        return self


class ParameterUpdateResponse(BaseModel):
    """§11.3 응답 항목. DB 컬럼 parameter_name 을 param_name 으로 노출."""
    model_config = ConfigDict(extra="forbid")

    id: UUID
    calibration_run_id: UUID
    target_type: str
    target_id: UUID
    param_name: str
    old_value_json: Any | None = None
    new_value_json: Any | None = None
    created_at: datetime
